File: hospital-ai/frontend/src/types/knowledgeGraph.ts
/** Intake Stage 6 — Patient Knowledge Graph types. */

export interface GraphNode {
  id: string;
  type: string;
  label: string;
  properties?: Record<string, unknown>;
}

export interface GraphRelationship {
  id: string;
  type: string;
  source: string;
  target: string;
  properties?: Record<string, unknown>;
}

export interface GraphStatistics {
  total_nodes?: number;
  total_relationships?: number;
  diseases?: number;
  symptoms?: number;
  medications?: number;
  doctors?: number;
  reports?: number;
  hospitals?: number;
  appointments?: number;
  procedures?: number;
  allergies?: number;
  vitals?: number;
  lab_tests?: number;
  risk_factors?: number;
  [key: string]: number | undefined;
}

export interface PatientGraphSummary {
  patient_name?: string;
  age?: number | null;
  conditions?: string[];
  current_medications?: string[];
  allergies?: string[];
  recent_procedures?: string[];
  risk_level?: string | null;
  recent_visits?: string[];
  graph_statistics?: GraphStatistics;
}

export interface KnowledgeGraph {
  id: string;
  processing_job_id?: string | null;
  patient_id: string;
  nodes_json: GraphNode[];
  relationships_json: GraphRelationship[];
  statistics_json: GraphStatistics;
  patient_summary_json: PatientGraphSummary;
  summary?: string | null;
  graph_version: number;
  node_count: number;
  relationship_count: number;
  builder_logs_json?: unknown[];
  validation_json?: Record<string, unknown>;
  status: string;
  error_message?: string | null;
  processing_time_ms?: number | null;
  created_at: string;
  updated_at?: string | null;
}

export interface KnowledgeGraphStartResult {
  job_id: string;
  patient_id: string;
  status: string;
  current_stage: string;
  next_stage: string;
  intake_completed: boolean;
  node_count: number;
  relationship_count: number;
  graph_version: number;
  processing_time_ms: number;
  summary: string;
  statistics: GraphStatistics;
  patient_summary: PatientGraphSummary;
  nodes: GraphNode[];
  relationships: GraphRelationship[];
  warnings: string[];
  knowledge_graph: KnowledgeGraph;
  processing_job: { id: string; current_stage: string; status: string };
  patient_context_version: number;
}

export const NODE_COLORS: Record<string, string> = {
  Patient: '#0f766e',
  Disease: '#dc2626',
  Symptom: '#ea580c',
  Medication: '#2563eb',
  Allergy: '#ca8a04',
  Doctor: '#0891b2',
  Hospital: '#4f46e5',
  Department: '#6366f1',
  Appointment: '#7c3aed',
  Procedure: '#9333ea',
  LabTest: '#a21caf',
  VitalSign: '#db2777',
  MedicalReport: '#57534e',
  MedicalRecord: '#57534e',
  RiskFactor: '#b91c1c',
  Insurance: '#0d9488',
  EmergencyContact: '#64748b',
};

File: hospital-ai/backend/app/schemas/knowledge_graph.py
"""Pydantic schemas for Intake Stage 6 — Patient Knowledge Graph."""

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.registration import ProcessingJobOut


class KnowledgeGraphStartRequest(BaseModel):
    job_id: UUID


class GraphNodeOut(BaseModel):
    id: str
    type: str
    label: str
    properties: Dict[str, Any] = Field(default_factory=dict)


class GraphRelationshipOut(BaseModel):
    id: str
    type: str
    source: str
    target: str
    properties: Dict[str, Any] = Field(default_factory=dict)


class KnowledgeGraphOut(BaseModel):
    id: UUID
    processing_job_id: Optional[UUID] = None
    patient_id: UUID
    nodes_json: List[Any] = Field(default_factory=list)
    relationships_json: List[Any] = Field(default_factory=list)
    statistics_json: Dict[str, Any] = Field(default_factory=dict)
    patient_summary_json: Dict[str, Any] = Field(default_factory=dict)
    summary: Optional[str] = None
    graph_version: int = 1
    node_count: int = 0
    relationship_count: int = 0
    builder_logs_json: List[Any] = Field(default_factory=list)
    validation_json: Dict[str, Any] = Field(default_factory=dict)
    status: str = "Completed"
    error_message: Optional[str] = None
    processing_time_ms: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


class KnowledgeGraphStartResponse(BaseModel):
    job_id: UUID
    patient_id: UUID
    status: str
    current_stage: str = "Completed"
    next_stage: str = "Completed"
    intake_completed: bool = True
    node_count: int
    relationship_count: int
    graph_version: int
    processing_time_ms: int
    summary: str
    statistics: Dict[str, Any] = Field(default_factory=dict)
    patient_summary: Dict[str, Any] = Field(default_factory=dict)
    nodes: List[GraphNodeOut] = Field(default_factory=list)
    relationships: List[GraphRelationshipOut] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)
    knowledge_graph: KnowledgeGraphOut
    processing_job: ProcessingJobOut
    patient_context_version: int = 1


class KnowledgeGraphStatusResponse(BaseModel):
    job_id: UUID
    status: str
    current_stage: str
    next_stage: str
    progress_pct: int
    kg_completed: bool = False
    intake_completed: bool = False
    node_count: Optional[int] = None
    relationship_count: Optional[int] = None
    error_message: Optional[str] = None
    processing_job: ProcessingJobOut
